jitrebalance/jitrebalance.py

```diff
@@ -0,0 +1,160 @@
+    funder = chan['msatoshi_to_us_max'] == chan['msatoshi_total']
+    forward_amt = Millisatoshi(onion['per_hop_v0']['forward_amount'])
+    incoming_amt = Millisatoshi(htlc['amount'])
+    fee = incoming_amt - forward_amt
+    pprint(fee)
+    # Compute the amount we need to rebalance, give us a bit of breathing room
+    # while we're at it (25% more rebalancing than strictly necessary) so we
+    # don't end up with a completely unbalanced channel right away again, and
+    # to account for a bit of fuzziness when it comes to dipping into the
+    # reserve.
+    amt = ceil(int(forward_amt - chan['spendable_msat']) * 1.25)
+
+    # If we have a higher balance than is required we don't need to rebalance,
+    # just stop here.
+    if amt <= 0:
+        request.set_result({"result": "continue"})
+        return
+
+    # Compute the last leg of the route first, so we know the parameters to
+    # traverse that last edge.
+    reverse_chan = plugin.rpc.listchannels(scid)['channels']
+    assert(len(reverse_chan) == 2)
+    reverse_chan = [
+        c for c in reverse_chan if c['channel_flags'] != chan['direction']
+    ][0]
+
+    if reverse_chan is None:
+        print("Could not compute parameters for the last hop")
+        request.set_result({"result": "continue"})
+        return
+
+    last_amt = ceil(float(amt) +
+                    float(amt) * reverse_chan['fee_per_millionth'] / 10**6 +
+                    reverse_chan['base_fee_millisatoshi'])
+    last_cltv = 9 + reverse_chan['delay']
+
+    # Exclude the channel we are trying to rebalance when searching for a
+    # path. We will manually append it to the route and bump the other
+    # parameters so it can be used afterwards
+    exclusions = [
+        "{scid}/{direction}".format(scid=scid, direction=chan['direction'])
+    ]
+
+    route = plugin.rpc.getroute(
+        node_id=peer['id'],
+        msatoshi=last_amt,
+        riskfactor=1,
+        exclude=exclusions,
+        cltv=last_cltv,
+    )['route']
+
+    # Append the last hop we computed manually above
+    route += [{
+        'id': plugin.node_id,
+        'channel': scid,
+        'direction': chan['direction'],
+        'msatoshi': amt,
+        'amount_msat': '{}msat'.format(amt),
+        'delay': 9
+    }]
+
+    # We're about to initiate a rebalancing, we'd better remember how we can
+    # settle it once we see it back here.
+
+    payment_key = secrets.token_bytes(32)
+    payment_hash = hashlib.sha256(payment_key).hexdigest()
+    plugin.rebalances[payment_hash] = {
+        "payment_key": binascii.hexlify(payment_key).decode('ASCII'),
+        "payment_hash": payment_hash,
+        "request": request,
+    }
+
+    # After all this work we're finally in a position to judge whether a
+    # rebalancing is worth it at all. The rebalancing is considered worth it
+    # if the fees we're about to pay are less than or equal to the fees we get
+    # out of forwarding the payment.
```

Comment:
But we are not doing this decision yet, or do we?

Just thinking:
Additionally, even if we do this correctly, a mean node might know about our JIT rebalance feature node enough so he will get us to 'drain' our other 'cheap enough' channels, leaving all the capacity with the higher fee channels. If he controls these channels, he can then readjust the fees to force us overpaying the next time we want to use our channels, right?

Reply:
> a mean node might know about our JIT rebalance feature node enough

Hmm .. Interesting but how would it ? If you don't disclose it in your alias that you use jitrouting.

>  If he controls these channels

It would have to control most of our channels, then ? Both ones to drain and ones used to rebalance. Here the Sybil att would have a certain cost if you are a minimum well connected ?

Reply:
I know this is all very hypothetical and complicated effort for steeling on fees, and I'm not even sure if this es economically viable at least since the victim can choose to close the extorted channel and the attacker has costs upfront on routing and such, but its a thought experiment:

1. Figuring out if a node uses JIT rebalancing is rather easy, at least by sending two circular payments that would not fit both through the same channel at the vitctim. if it works, its likely the node uses some rebalancing algo. There may also be other options, where the attacker has a testing channel with very low fees to the victim to bias rebalancing through this testchannel, ...

2. Sybil attacking this is also not hard, as the attacker overpays on fees until all the victims honest channels are empty, except his channels and one would be enough in theory.